File: Source/AlphaTab.JavaScript/Platform/JavaScript/AlphaSynthWebWorkerApi.cs
﻿using System;
using AlphaTab.Audio.Synth;
using AlphaTab.Audio.Synth.Midi;
using AlphaTab.Audio.Synth.Util;
using AlphaTab.Haxe.Js.Html;
using AlphaTab.Model;
using AlphaTab.Util;
using Haxe;
using Haxe.Js.Html;
using Phase;

namespace AlphaTab.Platform.JavaScript
{
    /// <summary>
    /// a WebWorker based alphaSynth which uses the given player as output.
    /// </summary>
    internal class AlphaSynthWebWorkerApi : IAlphaSynth
    {
        private readonly Worker _synth;
        private readonly ISynthOutput _output;

        private bool _workerIsReadyForPlayback;
        private bool _workerIsReady;
        private bool _outputIsReady;
        private PlayerState _state;
        private float _masterVolume;
        private float _metronomeVolume;
        private double _playbackSpeed;
        private int _tickPosition;
        private double _timePosition;
        private bool _isLooping;
        private PlaybackRange _playbackRange;

        /// <inheritdoc />
        public bool IsReady => _workerIsReady && _outputIsReady;

        /// <inheritdoc />
        public bool IsReadyForPlayback => _workerIsReadyForPlayback;

        /// <inheritdoc />
        public PlayerState State => _state;

        /// <inheritdoc />
        public LogLevel LogLevel
        {
            get => Logger.LogLevel;
            set
            {
                Logger.LogLevel = value;
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdSetLogLevel,
                    value = value
                });
            }
        }

        /// <inheritdoc />
        public float MasterVolume
        {
            get => _masterVolume;
            set
            {
                value = SynthHelper.ClampF(value, SynthConstants.MinVolume, SynthConstants.MaxVolume);
                _masterVolume = value;
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdSetMasterVolume,
                    value = value
                });
            }
        }

        /// <inheritdoc />
        public float MetronomeVolume
        {
            get => _metronomeVolume;
            set
            {
                value = SynthHelper.ClampF(value, SynthConstants.MinVolume, SynthConstants.MaxVolume);
                _metronomeVolume = value;
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdSetMetronomeVolume,
                    value = value
                });
            }
        }

        /// <inheritdoc />
        public double PlaybackSpeed
        {
            get => _playbackSpeed;
            set
            {
                value = SynthHelper.ClampD(value, SynthConstants.MinPlaybackSpeed, SynthConstants.MaxPlaybackSpeed);
                _playbackSpeed = value;
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdSetPlaybackSpeed,
                    value = value
                });
            }
        }

        /// <inheritdoc />
        public int TickPosition
        {
            get => _tickPosition;
            set
            {
                if (value < 0)
                {
                    value = 0;
                }

                _tickPosition = value;
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdSetTickPosition,
                    value = value
                });
            }
        }

        /// <inheritdoc />
        public double TimePosition
        {
            get => _timePosition;
            set
            {
                if (value < 0)
                {
                    value = 0;
                }

                _timePosition = value;
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdSetTimePosition,
                    value = value
                });
            }
        }

        /// <inheritdoc />
        public bool IsLooping
        {
            get => _isLooping;
            set
            {
                _isLooping = value;
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdSetIsLooping,
                    value = value
                });
            }
        }

        /// <inheritdoc />
        public PlaybackRange PlaybackRange
        {
            get => _playbackRange;
            set
            {
                if (value != null)
                {
                    if (value.StartTick < 0)
                    {
                        value.StartTick = 0;
                    }

                    if (value.EndTick < 0)
                    {
                        value.EndTick = 0;
                    }
                }

                _playbackRange = value;
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdSetPlaybackRange,
                    value = value
                });
            }
        }

        public AlphaSynthWebWorkerApi(ISynthOutput player, string alphaSynthScriptFile, LogLevel logLevel)
        {
            _output = player;
            _output.Ready += OnOutputReady;
            _output.SamplesPlayed += OnOutputSamplesPlayed;
            _output.SampleRequest += OnOutputSampleRequest;
            _output.Finished += OnOutputFinished;

            _output.Open();

            try
            {
                HaxeString script = "importScripts('" + alphaSynthScriptFile + "')";
                var blob = new Blob(Script.Write<object>("[ script ]"));
                _synth = new Worker(URL.CreateObjectURL(blob));
            }
            catch
            {
                // fallback to direct worker
                try
                {
                    _synth = new Worker(alphaSynthScriptFile);
                }
                catch (Exception e)
                {
                    Logger.Error("AlphaSynth", "Failed to create WebWorker: " + e);
                    // TODO: fallback to synchronous mode
                }
            }

            _synth.AddEventListener("message", (Action<MessageEvent>)HandleWorkerMessage, false);

            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdInitialize,
                sampleRate = _output.SampleRate,
                logLevel = logLevel
            });

            MasterVolume = 1;
            PlaybackSpeed = 1;
            MetronomeVolume = 0;
        }

        public void Destroy()
        {
            _synth.Terminate();
        }

        //
        // API communicating with the web worker

        /// <inheritdoc />
        public bool Play()
        {
            if (State == PlayerState.Playing || !IsReadyForPlayback)
            {
                return false;
            }

            _output.Activate();
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdPlay
            });
            return true;
        }

        /// <inheritdoc />
        public void Pause()
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdPause
            });
        }

        /// <inheritdoc />
        public void PlayPause()
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdPlayPause
            });
        }

        /// <inheritdoc />
        public void Stop()
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdStop
            });
        }

        /// <inheritdoc />
        public void LoadSoundFont(byte[] data)
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdLoadSoundFontBytes,
                data = data
            });
        }

        public void LoadSoundFontFromUrl(string data, Action<ProgressEventArgs> progress)
        {
            var url = data.As<string>();
            Logger.Info("AlphaSynth", "Start loading Soundfont from url " + url);
            var request = new XMLHttpRequest();
            request.Open("GET", url, true);
            request.ResponseType = XMLHttpRequestResponseType.ARRAYBUFFER;
            request.OnLoad = (Action<Event>)(e =>
            {
                var buffer = new Uint8Array(request.Response);
                _synth.PostMessage(new
                {
                    cmd = AlphaSynthWebWorker.CmdLoadSoundFontBytes,
                    data = buffer.As<byte[]>()
                });
            });
            request.OnError = (Action<Event>)(e =>
            {
                Logger.Error("AlphaSynth", "Loading failed: " + e.Member<string>("message"));
                OnSoundFontLoadFailed(e.As<Exception>());
            });
            request.OnProgress = (Action<Event>)(e =>
            {
                Logger.Debug("AlphaSynth",
                    "Soundfont downloading: " + e.Member<int>("loaded") + "/" + e.Member<int>("total") + " bytes");
                progress(new ProgressEventArgs(e.Member<int>("loaded"), e.Member<int>("total")));
            });
            request.Send();
        }

        public void LoadMidiFile(MidiFile midi)
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdLoadMidi,
                midi = JsonConverter.MidiFileToJsObject(midi)
            });
        }

        /// <inheritdoc />
        public void SetChannelMute(int channel, bool mute)
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdSetChannelMute,
                channel = channel,
                mute = mute
            });
        }

        /// <inheritdoc />
        public void ResetChannelStates()
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdResetChannelStates
            });
        }

        /// <inheritdoc />
        public void SetChannelSolo(int channel, bool solo)
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdSetChannelSolo,
                channel = channel,
                solo = solo
            });
        }

        /// <inheritdoc />
        public void SetChannelVolume(int channel, float volume)
        {
            volume = SynthHelper.ClampF(volume, SynthConstants.MinVolume, SynthConstants.MaxVolume);
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdSetChannelVolume,
                channel = channel,
                volume = volume
            });
        }

        /// <inheritdoc />
        public void SetChannelProgram(int channel, byte program)
        {
            program = SynthHelper.ClampB(program, SynthConstants.MinProgram, SynthConstants.MaxProgram);
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWebWorker.CmdSetChannelProgram,
                channel = channel,
                program = program
            });
        }

        public virtual void HandleWorkerMessage(MessageEvent e)
        {
            var data = e.Data;
            string cmd = data.cmd;
            switch (cmd)
            {
                case AlphaSynthWebWorker.CmdReady:
                    _workerIsReady = true;
                    CheckReady();
                    break;
                case AlphaSynthWebWorker.CmdReadyForPlayback:
                    _workerIsReadyForPlayback = true;
                    CheckReadyForPlayback();
                    break;
                case AlphaSynthWebWorker.CmdPositionChanged:
                    _timePosition = data.currentTime;
                    _tickPosition = data.currentTick;
                    OnPositionChanged(new PositionChangedEventArgs(data.currentTime,
                        data.endTime,
                        data.currentTick,
                        data.endTick));
                    break;
                case AlphaSynthWebWorker.CmdPlayerStateChanged:
                    _state = data.state;
                    OnStateChanged(new PlayerStateChangedEventArgs(data.state, data.stopped));
                    break;
                case AlphaSynthWebWorker.CmdFinished:
                    OnFinished();
                    break;
                case AlphaSynthWebWorker.CmdSoundFontLoaded:
                    OnSoundFontLoaded();
                    break;
                case AlphaSynthWebWorker.CmdSoundFontLoadFailed:
                    OnSoundFontLoadFailed(data.error);
                    break;
                case AlphaSynthWebWorker.CmdMidiLoaded:
                    CheckReadyForPlayback();
                    OnMidiLoaded();
                    break;
                case AlphaSynthWebWorker.CmdMidiLoadFailed:
                    CheckReadyForPlayback();
                    OnMidiLoadFailed(data.error);
                    break;
                case AlphaSynthWebWorker.CmdLog:
                    Logger.Log(data.level, "AlphaSynth", data.message);
                    break;

                // output communication ( output <- worker )
                case AlphaSynthWorkerSynthOutput.CmdOutputSequencerFinished:
                    _output.SequencerFinished();
                    break;
                case AlphaSynthWorkerSynthOutput.CmdOutputAddSamples:
                    _output.AddSamples(data.samples);
                    break;
                case AlphaSynthWorkerSynthOutput.CmdOutputPlay:
                    _output.Play();
                    break;
                case AlphaSynthWorkerSynthOutput.CmdOutputPause:
                    _output.Pause();
                    break;
                case AlphaSynthWorkerSynthOutput.CmdOutputResetSamples:
                    _output.ResetSamples();
                    break;
            }
        }

        private void CheckReady()
        {
            if (IsReady)
            {
                OnReady();
            }
        }


        private void CheckReadyForPlayback()
        {
            if (IsReadyForPlayback)
            {
                OnReadyForPlayback();
            }
        }

        #region Events

        public event Action Ready;

        protected virtual void OnReady()
        {
            var handler = Ready;
            if (handler != null)
            {
                handler();
            }
        }

        public event Action ReadyForPlayback;

        protected virtual void OnReadyForPlayback()
        {
            var handler = ReadyForPlayback;
            if (handler != null)
            {
                handler();
            }
        }

        public event Action Finished;

        protected virtual void OnFinished()
        {
            var handler = Finished;
            if (handler != null)
            {
                handler();
            }
        }

        public event Action SoundFontLoaded;

        protected virtual void OnSoundFontLoaded()
        {
            var handler = SoundFontLoaded;
            if (handler != null)
            {
                handler();
            }
        }

        public event Action<Exception> SoundFontLoadFailed;

        protected virtual void OnSoundFontLoadFailed(Exception e)
        {
            var handler = SoundFontLoadFailed;
            if (handler != null)
            {
                handler(e);
            }
        }


        public event Action MidiLoaded;

        protected virtual void OnMidiLoaded()
        {
            var handler = MidiLoaded;
            if (handler != null)
            {
                handler();
            }
        }

        public event Action<Exception> MidiLoadFailed;

        protected virtual void OnMidiLoadFailed(Exception e)
        {
            var handler = MidiLoadFailed;
            if (handler != null)
            {
                handler(e);
            }
        }

        public event Action<PlayerStateChangedEventArgs> StateChanged;

        protected virtual void OnStateChanged(PlayerStateChangedEventArgs e)
        {
            var handler = StateChanged;
            if (handler != null)
            {
                handler(e);
            }
        }


        public event Action<PositionChangedEventArgs> PositionChanged;

        protected virtual void OnPositionChanged(PositionChangedEventArgs e)
        {
            var handler = PositionChanged;
            if (handler != null)
            {
                handler(e);
            }
        }

        #endregion

        //
        // output communication ( output -> worker )


        public void OnOutputSampleRequest()
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWorkerSynthOutput.CmdOutputSampleRequest
            });
        }

        public void OnOutputFinished()
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWorkerSynthOutput.CmdOutputFinished
            });
        }

        public void OnOutputSamplesPlayed(int samples)
        {
            _synth.PostMessage(new
            {
                cmd = AlphaSynthWorkerSynthOutput.CmdOutputSamplesPlayed,
                samples = samples
            });
        }

        private void OnOutputReady()
        {
            _outputIsReady = true;
            CheckReady();
        }
    }
}

File: packages/alphatab/src/platform/worker/AlphaSynthWebWorkerApi.ts
import { Environment } from '@coderline/alphatab/Environment';
import {
    EventEmitter,
    EventEmitterOfT,
    type IEventEmitter,
    type IEventEmitterOfT
} from '@coderline/alphatab/EventEmitter';
import { Logger } from '@coderline/alphatab/Logger';
import type { LogLevel } from '@coderline/alphatab/LogLevel';
import type { MidiEventType } from '@coderline/alphatab/midi/MidiEvent';
import type { MidiFile } from '@coderline/alphatab/midi/MidiFile';
import { JsonConverter } from '@coderline/alphatab/model/JsonConverter';
import { ModelUtils } from '@coderline/alphatab/model/ModelUtils';
import type { Score } from '@coderline/alphatab/model/Score';
import type {
    IAlphaSynthWorker,
    IAlphaSynthWorkerMessage
} from '@coderline/alphatab/platform/worker/AlphaTabWorkerProtocol';
import type { Settings } from '@coderline/alphatab/Settings';
import type { BackingTrackSyncPoint, IAlphaSynth } from '@coderline/alphatab/synth/IAlphaSynth';
import type { ISynthOutput } from '@coderline/alphatab/synth/ISynthOutput';
import { MidiEventsPlayedEventArgs } from '@coderline/alphatab/synth/MidiEventsPlayedEventArgs';
import type { PlaybackRange } from '@coderline/alphatab/synth/PlaybackRange';
import { PlaybackRangeChangedEventArgs } from '@coderline/alphatab/synth/PlaybackRangeChangedEventArgs';
import { PlayerState } from '@coderline/alphatab/synth/PlayerState';
import { PlayerStateChangedEventArgs } from '@coderline/alphatab/synth/PlayerStateChangedEventArgs';
import { PositionChangedEventArgs } from '@coderline/alphatab/synth/PositionChangedEventArgs';
import { SynthConstants } from '@coderline/alphatab/synth/SynthConstants';

/**
 * a WebWorker based alphaSynth which uses the given player as output.
 * @internal
 */
export class AlphaSynthWebWorkerApi implements IAlphaSynth {
    private _synth!: IAlphaSynthWorker;
    private _output: ISynthOutput;
    private _workerIsReadyForPlayback: boolean = false;
    private _workerIsReady: boolean = false;
    private _outputIsReady: boolean = false;
    private _state: PlayerState = PlayerState.Paused;
    private _masterVolume: number = 0;
    private _metronomeVolume: number = 0;
    private _countInVolume: number = 0;
    private _playbackSpeed: number = 0;
    private _isLooping: boolean = false;
    private _playbackRange: PlaybackRange | null = null;
    private _midiEventsPlayedFilter: MidiEventType[] = [];
    private _loadedMidiInfo?: PositionChangedEventArgs;
    private _currentPosition: PositionChangedEventArgs = new PositionChangedEventArgs(0, 0, 0, 0, false, 120, 120);

    public get output(): ISynthOutput {
        return this._output;
    }

    public get isReady(): boolean {
        return this._workerIsReady && this._outputIsReady;
    }

    public get isReadyForPlayback(): boolean {
        return this._workerIsReadyForPlayback;
    }

    public get state(): PlayerState {
        return this._state;
    }

    public get logLevel(): LogLevel {
        return Logger.logLevel;
    }

    public get worker(): IAlphaSynthWorker {
        return this._synth;
    }

    public set logLevel(value: LogLevel) {
        Logger.logLevel = value;
        this._synth.postMessage({
            cmd: 'alphaSynth.setLogLevel',
            value: value
        });
    }

    public get masterVolume(): number {
        return this._masterVolume;
    }

    public set masterVolume(value: number) {
        value = Math.max(value, SynthConstants.MinVolume);
        this._masterVolume = value;
        this._synth.postMessage({
            cmd: 'alphaSynth.setMasterVolume',
            value: value
        });
    }

    public get metronomeVolume(): number {
        return this._metronomeVolume;
    }

    public set metronomeVolume(value: number) {
        value = Math.max(value, SynthConstants.MinVolume);
        this._metronomeVolume = value;
        this._synth.postMessage({
            cmd: 'alphaSynth.setMetronomeVolume',
            value: value
        });
    }
    public get countInVolume(): number {
        return this._countInVolume;
    }

    public set countInVolume(value: number) {
        value = Math.max(value, SynthConstants.MinVolume);
        this._countInVolume = value;
        this._synth.postMessage({
            cmd: 'alphaSynth.setCountInVolume',
            value: value
        });
    }

    public get midiEventsPlayedFilter(): MidiEventType[] {
        return this._midiEventsPlayedFilter;
    }

    public set midiEventsPlayedFilter(value: MidiEventType[]) {
        this._midiEventsPlayedFilter = value;
        this._synth.postMessage({
            cmd: 'alphaSynth.setMidiEventsPlayedFilter',
            value: Environment.prepareForPostMessage(value)
        });
    }

    public get playbackSpeed(): number {
        return this._playbackSpeed;
    }

    public set playbackSpeed(value: number) {
        value = ModelUtils.clamp(value, SynthConstants.MinPlaybackSpeed, SynthConstants.MaxPlaybackSpeed);
        this._playbackSpeed = value;
        this._synth.postMessage({
            cmd: 'alphaSynth.setPlaybackSpeed',
            value: value
        });
    }

    public get loadedMidiInfo(): PositionChangedEventArgs | undefined {
        return this.loadedMidiInfo;
    }

    public get currentPosition(): PositionChangedEventArgs {
        return this._currentPosition;
    }

    public get tickPosition(): number {
        return this._currentPosition.currentTick;
    }

    public set tickPosition(value: number) {
        if (value < 0) {
            value = 0;
        }
        this._currentPosition = new PositionChangedEventArgs(
            this._currentPosition.currentTime,
            this._currentPosition.endTime,
            value,
            this._currentPosition.endTick,
            true,
            this._currentPosition.originalTempo,
            this._currentPosition.modifiedTempo
        );
        this._synth.postMessage({
            cmd: 'alphaSynth.setTickPosition',
            value: value
        });
    }

    public get timePosition(): number {
        return this._currentPosition.currentTime;
    }

    public set timePosition(value: number) {
        if (value < 0) {
            value = 0;
        }
        this._currentPosition = new PositionChangedEventArgs(
            value,
            this._currentPosition.endTime,
            this._currentPosition.currentTick,
            this._currentPosition.endTick,
            true,
            this._currentPosition.originalTempo,
            this._currentPosition.modifiedTempo
        );
        this._synth.postMessage({
            cmd: 'alphaSynth.setTimePosition',
            value: value
        });
    }

    public get isLooping(): boolean {
        return this._isLooping;
    }

    public set isLooping(value: boolean) {
        this._isLooping = value;
        this._synth.postMessage({
            cmd: 'alphaSynth.setIsLooping',
            value: value
        });
    }

    public get playbackRange(): PlaybackRange | null {
        return this._playbackRange;
    }

    public set playbackRange(value: PlaybackRange | null) {
        if (value) {
            if (value.startTick < 0) {
                value.startTick = 0;
            }
            if (value.endTick < 0) {
                value.endTick = 0;
            }
        }
        this._playbackRange = value;
        this._synth.postMessage({
            cmd: 'alphaSynth.setPlaybackRange',
            value: Environment.prepareForPostMessage(value)
        });
    }

    public constructor(player: ISynthOutput, settings: Settings, synthWorker: IAlphaSynthWorker) {
        this._workerIsReadyForPlayback = false;
        this._workerIsReady = false;
        this._outputIsReady = false;
        this._state = PlayerState.Paused;
        this._masterVolume = 0.0;
        this._metronomeVolume = 0.0;
        this._playbackSpeed = 0.0;
        this._isLooping = false;
        this._playbackRange = null;
        this._output = player;
        this._output.ready.on(this._onOutputReady.bind(this));
        this._output.samplesPlayed.on(this.onOutputSamplesPlayed.bind(this));
        this._output.sampleRequest.on(this.onOutputSampleRequest.bind(this));
        this._output.open(settings.player.bufferTimeInMilliseconds);
        this._synth = synthWorker;
        this._synth.addEventListener('message', e => this.handleWorkerMessage(e));
        this._synth.postMessage({
            cmd: 'alphaSynth.initialize',
            sampleRate: this._output.sampleRate,
            logLevel: settings.core.logLevel,
            bufferTimeInMilliseconds: settings.player.bufferTimeInMilliseconds
        });
        this.masterVolume = 1;
        this.playbackSpeed = 1;
        this.metronomeVolume = 0;
    }

    public destroy(): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.destroy'
        });
    }

    //
    // API communicating with the web worker
    public play(): boolean {
        this._output.activate();
        this._synth.postMessage({
            cmd: 'alphaSynth.play'
        });
        return true;
    }

    public pause(): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.pause'
        });
    }

    public playPause(): void {
        this._output.activate();
        this._synth.postMessage({
            cmd: 'alphaSynth.playPause'
        });
    }

    public stop(): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.stop'
        });
    }

    public playOneTimeMidiFile(midi: MidiFile): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.playOneTimeMidiFile',
            midi: JsonConverter.midiFileToJsObject(Environment.prepareForPostMessage(midi))
        });
    }

    public loadSoundFont(data: Uint8Array, append: boolean): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.loadSoundFontBytes',
            data: Environment.prepareForPostMessage(data),
            append: append
        });
    }

    public resetSoundFonts(): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.resetSoundFonts'
        });
    }

    public loadMidiFile(midi: MidiFile): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.loadMidi',
            midi: JsonConverter.midiFileToJsObject(Environment.prepareForPostMessage(midi))
        });
    }

    public applyTranspositionPitches(transpositionPitches: Map<number, number>): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.applyTranspositionPitches',
            transpositionPitches: Environment.prepareForPostMessage(transpositionPitches)
        });
    }

    public setChannelTranspositionPitch(channel: number, semitones: number): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.setChannelTranspositionPitch',
            channel: channel,
            semitones: semitones
        });
    }

    public setChannelMute(channel: number, mute: boolean): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.setChannelMute',
            channel: channel,
            mute: mute
        });
    }

    public resetChannelStates(): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.resetChannelStates'
        });
    }

    public setChannelSolo(channel: number, solo: boolean): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.setChannelSolo',
            channel: channel,
            solo: solo
        });
    }

    public setChannelVolume(channel: number, volume: number): void {
        volume = Math.max(volume, SynthConstants.MinVolume);
        this._synth.postMessage({
            cmd: 'alphaSynth.setChannelVolume',
            channel: channel,
            volume: volume
        });
    }

    public handleWorkerMessage(e: MessageEvent<IAlphaSynthWorkerMessage>): void {
        const data = e.data;
        switch (data.cmd) {
            case 'alphaSynth.ready':
                this._workerIsReady = true;
                this._checkReady();
                break;
            case 'alphaSynth.destroyed':
                this._synth.terminate();
                break;
            case 'alphaSynth.readyForPlayback':
                this._workerIsReadyForPlayback = true;
                this._checkReadyForPlayback();
                break;
            case 'alphaSynth.positionChanged':
                this._currentPosition = data.args;
                (this.positionChanged as EventEmitterOfT<PositionChangedEventArgs>).trigger(this._currentPosition);
                break;
            case 'alphaSynth.midiEventsPlayed':
                (this.midiEventsPlayed as EventEmitterOfT<MidiEventsPlayedEventArgs>).trigger(
                    new MidiEventsPlayedEventArgs(data.events.map(JsonConverter.jsObjectToMidiEvent))
                );
                break;
            case 'alphaSynth.playerStateChanged':
                this._state = data.state;
                (this.stateChanged as EventEmitterOfT<PlayerStateChangedEventArgs>).trigger(
                    new PlayerStateChangedEventArgs(data.state, data.stopped)
                );
                break;
            case 'alphaSynth.playbackRangeChanged':
                this._playbackRange = data.playbackRange;
                (this.playbackRangeChanged as EventEmitterOfT<PlaybackRangeChangedEventArgs>).trigger(
                    new PlaybackRangeChangedEventArgs(this._playbackRange)
                );
                break;
            case 'alphaSynth.finished':
                (this.finished as EventEmitter).trigger();
                break;
            case 'alphaSynth.soundFontLoaded':
                (this.soundFontLoaded as EventEmitter).trigger();
                break;
            case 'alphaSynth.soundFontLoadFailed':
                (this.soundFontLoadFailed as EventEmitterOfT<Error>).trigger(data.error);
                break;
            case 'alphaSynth.midiLoaded':
                this._checkReadyForPlayback();
                this._loadedMidiInfo = data.args;
                (this.midiLoaded as EventEmitterOfT<PositionChangedEventArgs>).trigger(this._loadedMidiInfo);
                break;
            case 'alphaSynth.midiLoadFailed':
                this._checkReadyForPlayback();
                (this.midiLoadFailed as EventEmitterOfT<Error>).trigger(data.error);
                break;
            case 'alphaSynth.output.addSamples':
                this._output.addSamples(data.samples);
                break;
            case 'alphaSynth.output.play':
                this._output.play();
                break;
            case 'alphaSynth.output.pause':
                this._output.pause();
                break;
            case 'alphaSynth.output.destroy':
                this._output.destroy();
                break;
            case 'alphaSynth.output.resetSamples':
                this._output.resetSamples();
                break;
        }
    }

    private _checkReady(): void {
        if (this.isReady) {
            (this.ready as EventEmitter).trigger();
        }
    }

    private _checkReadyForPlayback(): void {
        if (this.isReadyForPlayback) {
            (this.readyForPlayback as EventEmitter).trigger();
        }
    }

    readonly ready: IEventEmitter = new EventEmitter();
    readonly readyForPlayback: IEventEmitter = new EventEmitter();
    readonly finished: IEventEmitter = new EventEmitter();
    readonly soundFontLoaded: IEventEmitter = new EventEmitter();
    readonly soundFontLoadFailed: IEventEmitterOfT<Error> = new EventEmitterOfT<Error>();
    readonly midiLoaded: IEventEmitterOfT<PositionChangedEventArgs> = new EventEmitterOfT<PositionChangedEventArgs>();
    readonly midiLoadFailed: IEventEmitterOfT<Error> = new EventEmitterOfT<Error>();
    readonly stateChanged: IEventEmitterOfT<PlayerStateChangedEventArgs> =
        new EventEmitterOfT<PlayerStateChangedEventArgs>();
    readonly positionChanged: IEventEmitterOfT<PositionChangedEventArgs> =
        new EventEmitterOfT<PositionChangedEventArgs>();
    readonly midiEventsPlayed: IEventEmitterOfT<MidiEventsPlayedEventArgs> =
        new EventEmitterOfT<MidiEventsPlayedEventArgs>();
    readonly playbackRangeChanged: IEventEmitterOfT<PlaybackRangeChangedEventArgs> =
        new EventEmitterOfT<PlaybackRangeChangedEventArgs>();

    //
    // output communication ( output -> worker )
    public onOutputSampleRequest(): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.output.sampleRequest'
        });
    }
    public onOutputSamplesPlayed(samples: number): void {
        this._synth.postMessage({
            cmd: 'alphaSynth.output.samplesPlayed',
            samples: samples
        });
    }

    private _onOutputReady(): void {
        this._outputIsReady = true;
        this._checkReady();
    }

    public loadBackingTrack(_score: Score): void {
        // ignore
    }

    public updateSyncPoints(_syncPoints: BackingTrackSyncPoint[]): void {
        // ignore
    }
}
